### PR TITLE
Updated: Azure Cognitive Search Plugin/ Free AI APIs 🔎

### PR DESCRIPTION
**Pull request by itzraiss**
## Summary

This pull request updates Azure Cognitive Search to the new name given by Microsoft: Azure AI Search, in all files where it is present. I also updated @azure/search-documents from version 11.3.2 to 12.0.0. I updated the default search API version from 2020-06-30 to 2023-11-01 in both api/package.json and package-lock.json. Finally, I created a tutorial with a simple step-by-step guide on how to use the plugin and simplified what was already there.

## Change Type

- [x] New feature how semantic search and search vetorial
- [x] This change requires a documentation update
- [x] Documentation update 

## Testing

All working 😁

### **Test Configuration**:

After the update, I performed all the tests, and everything went smoothly as expected.

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I conducted a self-review of the code at @zentooo.
- [x] I have commented in any complex areas of my code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes
- [x] Any changes dependent on mine have been merged and published in downstream modules.

### CODE REVIEW

**Comment by danny-avila** [comment]
Thanks @itzraiss 

I noticed in the .env.example and the guide, the environment variables were changed back to the old values. We should promote the new ones to eventually deprecate the old.

**Comment by itzraiss** [comment]
No problem, here is the updated code. (Sorry for the inconvenience, I'm still new to programming). 🙂

File: api/app/clients/tools/AzureAiSearch.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
+const { Tool } = require('langchain/tools');
+const { SearchClient, AzureKeyCredential } = require('@azure/search-documents');
+
+class AzureAISearch extends Tool {
+  constructor(fields = {}) {
+    super();
+    this.serviceEndpoint =
```
Comment (danny-avila) [inline api/app/clients/tools/AzureAiSearch.js]: I want this to be a bit cleaner here, can you test the following (from GPT)?


---
To improve the overall cleanliness and readability of your `AzureAISearch` class, here are several recommendations:

1. **Refactor Repeated Code**: Introduce a helper function to reduce redundancy when initializing properties. This function can streamline the process of checking and assigning values from various sources.

2. **Define Default Values as Constants**: For better code organization and readability, define default values as constants at the beginning of the class or in a separate configuration file.

3. **Simplify Getter Methods**: Simplify your getter methods by directly returning the value or default without unnecessary checks, as the checks are already handled in the constructor.

4. **Error Handling and Logging**: Improve error handling and logging for better diagnostics and maintenance.

5. **Use Async/Await Consistently**: Ensure consistency in the use of async/await for asynchronous operations.

6. **Documentation**: Enhance inline documentation for clarity and maintainability.

Here's a refactored version of your class incorporating these suggestions:

```javascript
const { Tool } = require('langchain/tools');
const { SearchClient, AzureKeyCredential } = require('@azure/search-documents');

class AzureAISearch extends Tool {
  static DEFAULT_API_VERSION = '2023-11-01';
  static DEFAULT_QUERY_TYPE = 'simple';
  static DEFAULT_TOP = 5;

  constructor(fields = {}) {
    super();
    this.initializeProperties(fields);
    this.initializeClient();
  }

  initializeProperties(fields) {
    const getValue = (fieldNames, defaultValue) => {
      for (const name of fieldNames) {
        const value = fields[name] || process.env[name];
        if (value !== undefined && value !== null) return value;
      }
      return defaultValue;
    };

    this.serviceEndpoint = getValue(['AZURE_AI_SEARCH_SERVICE_ENDPOINT', 'AZURE_COGNITIVE_SEARCH_SERVICE_ENDPOINT'], this.getServiceEndpoint());
    this.indexName = getValue(['AZURE_AI_SEARCH_INDEX_NAME', 'AZURE_COGNITIVE_SEARCH_INDEX_NAME'], this.getIndexName());
    this.apiKey = getValue(['AZURE_AI_SEARCH_API_KEY', 'AZURE_COGNITIVE_SEARCH_API_KEY'], this.getApiKey());
    this.apiVersion = getValue(['AZURE_AI_SEARCH_API_VERSION', 'AZURE_COGNITIVE_SEARCH_API_VERSION'], AzureAISearch.DEFAULT_API_VERSION);
    this.queryType = getValue(['AZURE_AI_SEARCH_SEARCH_OPTION_QUERY_TYPE', 'AZURE_COGNITIVE_SEARCH_SEARCH_OPTION_QUERY_TYPE'], AzureAISearch.DEFAULT_QUERY_TYPE);
    this.top = getValue(['AZURE_AI_SEARCH_SEARCH_OPTION_TOP', 'AZURE_COGNITIVE_SEARCH_SEARCH_OPTION_TOP'], AzureAISearch.DEFAULT_TOP);
    this.select = this.getSelect();
  }

  initializeClient() {
    this.client = new SearchClient(this.serviceEndpoint, this.indexName, new AzureKeyCredential(this.apiKey), { apiVersion: this.apiVersion });
  }

  // ... [Rest of your getter methods and _call method, refactored for simplicity and readability] ...

  async _call(query) {
    try {
      const searchOptions = {
        queryType: this.queryType,
        top: this.top,
        select: this.select
      };

      const searchResults = await this.client.search(query, searchOptions);
      return JSON.stringify(searchResults.results.map(result => result.document));
    } catch (error) {
      console.error(`Azure AI Search request failed: ${error}`);
      return 'There was an error with Azure AI Search.';
    }
  }
}

module.exports = AzureAISearch;
```

This refactored version provides a cleaner and more maintainable structure, with reduced redundancy and improved readability.

**Review by danny-avila — CHANGES_REQUESTED** [review]

**Comment by danny-avila** [comment]
> No problem, here is the updated code @danny-avila . (Sorry for the inconvenience, I'm still new to programming). 🙂

No worries, I dropped a suggestion from GPT to make it a little cleaner. I don't have access to this so I can't really test. After that, it looks good to me and we can merge!

File: api/app/clients/tools/AzureAiSearch.js (hunk 1 above)
Comment (itzraiss) [reply api/app/clients/tools/AzureAiSearch.js]: Alright @danny-avila , something like this?

Comment (danny-avila) [reply api/app/clients/tools/AzureAiSearch.js]: @itzraiss yes, thank you. over 40 lines of code removed from that change, thanks! (I assume it's working too)

**Review by danny-avila — APPROVED** [review]

**Pull request merged**